pkg/providers/cloudstack/cloudstack.go

```diff
@@ -1057,10 +1083,32 @@ func (p *cloudstackProvider) MachineConfigs(_ *cluster.Spec) []providers.Machine
 	return configs
 }
 
-func (p *cloudstackProvider) UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec) (bool, error) {
+func (p *cloudstackProvider) UpgradeNeeded(ctx context.Context, newSpec, currentSpec *cluster.Spec, cluster *types.Cluster, datacenterConfig providers.DatacenterConfig, machineConfigs []providers.MachineConfig) (bool, error) {
 	newV, oldV := newSpec.VersionsBundle.CloudStack, currentSpec.VersionsBundle.CloudStack
 
-	return newV.ClusterAPIController.ImageDigest != oldV.ClusterAPIController.ImageDigest, nil
+	if newV.ClusterAPIController.ImageDigest != oldV.ClusterAPIController.ImageDigest {
+		logger.V(3).Info("New provider controller is different from the old spec")
+		return true, nil
+	}
+	cc, err := p.providerKubectlClient.GetEksaCluster(ctx, cluster, newSpec.Cluster.Name)
```

Comment:
I missed this the first time, sorry!
You can get this cluster from `currentSpec.Cluster`
Same for the other providers

Reply:
using `currentSpec.Cluster` instead of querying with kubectl